### PR TITLE
Give exactly-zero results of remainder, floor_divide and divmod the correct sign

### LINKED ISSUE #40028: jnp.remainder gives a zero result the sign of the dividend, not the divisor
### Description

The `jnp.remainder` docstring says the result has "same sign as the elements of `x2`". For a zero result it has the sign of `x1` instead.

```python
import jax, jax.numpy as jnp, numpy as np
jax.config.update("jax_enable_x64", True)

a = jnp.array([-1.0, -4.0, 0.0, 4.0])
b = jnp.array([ 1.0,  2.0, -3.0, -2.0])

jnp.remainder(a, b)                        # [-0., -0.,  0.,  0.]
np.remainder(np.asarray(a), np.asarray(b)) # [ 0.,  0., -0., -0.]
```

All four disagree with the documented rule. The first two should be `+0.0` because `x2 > 0`, the last two `-0.0` because `x2 < 0`. NumPy and Python's `%` both give the documented answer.

`jnp.fmod` is correct on the same inputs, which is what localizes this. `fmod` is defined to take the dividend's sign, so the difference is only in the floored fixup.

### Cause

In `remainder` in `jax/_src/numpy/ufuncs.py`:

```python
trunc_mod = lax.rem(x1, x2)
trunc_mod_not_zero = lax.ne(trunc_mod, zero)
do_plus = lax.bitwise_and(
    lax.ne(lax.lt(trunc_mod, zero), lax.lt(x2, zero)), trunc_mod_not_zero)
out = lax.select(do_plus, lax.add(trunc_mod, x2), trunc_mod)
```

`lax.rem` follows the dividend. The fixup is guarded on `trunc_mod_not_zero`, which is needed for the value, since adding `x2` to a zero remainder would give `x2`. But it also skips the sign, so `lax.rem`'s zero survives.

`jnp.mod` and `jnp.divmod` inherit it. Integers are unaffected, having no signed zero.

This is not only cosmetic: `1 / result` is `inf` or `-inf` depending on the sign, so it propagates.

Not the subnormal flush-to-zero divergence from Sharp Bits. No input here is subnormal.

### Fix

Give an exact zero the divisor's sign, for floating dtypes only:

```python
out = lax.select(do_plus, lax.add(trunc_mod, x2), trunc_mod)
if dtypes.issubdtype(out.dtype, np.floating):
    out = _where(trunc_mod_not_zero, out, copysign(zero, x2))
```

Happy to send a PR with a test if you want it that way.

### System info

jax 0.11.0, jaxlib 0.11.0, CPU backend, macOS arm64, Python 3.12.

Same bug exists in two other array libraries, so it seems to be a common shape rather than a one-off: ml-explore/mlx#4315 (fixed) and pytorch/pytorch#193755.

### PR DESCRIPTION
**Pull request by kodlan**
Fixes #40028.

For floating point inputs, `jnp.remainder` returned an exactly-zero result with the sign of the dividend, while NumPy, Python's `%` and jax's own docstring say it should have the divisor's sign. `jnp.floor_divide` had the matching problem for a zero quotient (for example `-0.0 // 3.0` returned `0.0` instead of `-0.0`), and `jnp.mod`, `jnp.divmod` and the `%`, `//`, `divmod` operators inherit both. The sign of a zero is observable, `1 / result` flips between `inf` and `-inf`, so this propagates to real wrong values.

The cause is that `_float_divmod` in `jax/_src/numpy/ufuncs.py` is a port of CPython's `float_divmod` (the code comment points there) but drops its two zero-sign branches, `mod = copysign(0.0, wx)` and `floordiv = copysign(0.0, vx / wx)`. `remainder` is a separate formula with the same gap: its sign fixup is gated on the remainder being nonzero, so a zero keeps `lax.rem`'s dividend sign. The fix restores those branches with `copysign`, which is a no-op for every nonzero result (they already carry the right sign), so only the zeros change. Verified against NumPy over a special-value grid and a 200k-sample random sweep: no sign mismatches left and no nonzero value changes, on CPU and GPU, eager and jit, for float64/float32/float16/bfloat16 (the only remaining CPU-only divergence involves the smallest subnormals, where XLA:CPU comparisons flush to zero; that pre-exists this change and also affects the untouched `fmod`). Integer dtypes take a separate branch and are untouched. The copysign is skipped for sub-16-bit floats because `signbit` does not support them; float8 keeps its previous behavior, and a test pins that those ops keep working there.

This also lets us drop 15 entries from `tests/array_api_skips.txt` ("Returns wrong zero sign" section). I ran the pinned array-api-tests suite locally with the trimmed skips file and the full special-cases file passes. The two `__floordiv__` inf entries stay: there jax returns `-1.0`, matching NumPy and CPython against the spec, which is a separate (deliberate) divergence this PR does not touch.

One note for reviewers: the gradient of `remainder` changes at points where the result is exactly zero and the divisor is negative (the copysign flips the tangent sign there). That set has measure zero and sits on discontinuities of the primal where the derivative is a convention anyway; gradients everywhere else are bit-identical, and `floor_divide`'s gradient (zero via `round`) is unaffected.

### CODE REVIEW

**Review by CAOShurong — COMMENTED** [review]
Exact-head verification at `76684e99f` (base `6aa95e6c`, CPU jaxlib 0.11.1 / Windows, Python 3.13) — independently reproduced both directions, full operator suite green, plus an independent NumPy-equivalence sweep. AI assistance disclosure: code inspection, test execution and analysis were prepared with the help of an AI coding agent (Codex CLI); all results below were executed and checked locally by me.

**Pre-fix behavior confirmed on unmodified main `jnp` sources** (main's `jax/_src/numpy/ufuncs.py` over the PR tree):

- `remainder(-0.0, 5.0)` → `-0.0`; `remainder(0.0, -5.0)` → `+0.0` (NumPy: `+0.0` / `-0.0`)
- `floor_divide(-0.0, 5.0)` → `+0.0` (NumPy: `-0.0`)
- `divmod(-0.0, 5.0)` → `(+0.0, -0.0)` (NumPy: `(-0.0, +0.0)`)

so all three reported families reproduce, eager and under `jax.jit`.

**Post-fix at PR head:** all of the above match NumPy exactly, including the `jit` path (`remainder(-0.0, 5.0)` → `+0.0`). Also probed bfloat16 outside the test parametrization: `remainder(bfloat16(-0.0), bfloat16(5.0))` → `+0.0`, so the ≥16-bit `copysign` gate holds where it matters.

**Tests:** the 18 new parametrized cases pass (`test_float_zero_result_sign` ×10 across float16/32/64/bfloat16 × {remainder, mod, floor_divide, divmod} + `test_float8_division_ops` ×8), and the full `tests/lax_numpy_operators_test.py` is green at head: **1650 passed, 2 skipped**.

**Independent sweep:** 20 random trials per dtype (float32/float64, ~532k elementwise comparisons vs NumPy) with forced exact-zero dividends/divisors and magnitude scaling 1e-3…1e3, comparing value *and* signbit for `remainder`, `mod`, `floor_divide`, `divmod`: **zero mismatches** with x64 enabled. This is consistent with the copysign-no-op-for-nonzeros argument: floor-mod non-zero results already carry the divisor's sign, so only zeros can change.

**Edge cases checked:**

- The dtype gate keys off `x1.dtype` after `promote_args_numeric`, so mixed int/float inputs promote first and take the float path correctly; pure-integer calls skip the new code entirely (verified by reading the promote ordering).
- The two `__floordiv__(finite, ±infinity) -> -0` entries intentionally left in `array_api_skips.txt` are genuinely orthogonal: at head `floor_divide(3, -inf)` still returns `-1` (matching NumPy, diverging from the array-API special-case table), i.e. those skips pin a spec disagreement, not a regression of this change. Skip removal scope is exactly right.
- NaN propagation unaffected: `copysign` on a NaN result keeps it NaN; `_set_error_if_nan` still runs after the sign fixup.

Nice find on the CPython `float_divmod` port having dropped its two zero-sign branches — the diff restores exactly those semantics with minimal surface. LGTM.

File: tests/lax_numpy_operators_test.py [hunk 1]
```diff
@@ -724,6 +724,50 @@ def test_float_floor_divide_zero(self, dtype):
...
+      fn=["remainder", "mod", "floor_divide", "divmod"],
+  )
+  def test_float_zero_result_sign(self, fn, dtype):
+    # Regression test for https://github.com/jax-ml/jax/issues/40028: an
+    # exactly-zero result must match NumPy's sign (the divisor's sign for
+    # remainder/mod, the true quotient's sign for floor_divide).
+    vals = np.array([-4.0, -3.0, -1.0, -0.5, -0.0, 0.0, 0.5, 1.0, 3.0, 4.0,
+                     np.inf, -np.inf], dtype=dtype)
+    x1, x2 = map(np.ravel, np.meshgrid(vals, vals))
+    np_fun, jnp_fun = getattr(np, fn), getattr(jnp, fn)
```
Comment (jakevdp) [inline tests/lax_numpy_operators_test.py]: Seems like we could simplify this; maybe an approach like this?
```python
args_maker = lambda: [x1, x2]

def jnp_fun(*args):
  result = getattr(jnp, fn)(*args)
  return jax.tree.map(jnp.signbit, result)

def np_fun(*args):
  result = getattr(np, fn)(*args)
  return jax.tree.map(np.signbit, result)

self._CheckAgainstNumpy(np_fun, jnp_fun, args_maker)
self._CompileAndCheck(jnp_fun, args_maker)
```
It might need some adjustment for NaNs, but I think this is cleaner. What do you think?

**Review by jakevdp — CHANGES_REQUESTED** [review]

File: tests/lax_numpy_operators_test.py (hunk 1 above)
Comment (jakevdp) [reply tests/lax_numpy_operators_test.py]: Maybe `jax.tree.map(lambda x: np.isnan(x) | np.signbit(x), result)` to avoid checking signbit for NaN values.

File: tests/lax_numpy_operators_test.py [hunk 2]
```diff
@@ -724,6 +724,50 @@ def test_float_floor_divide_zero(self, dtype):
     self._CheckAgainstNumpy(np.floor_divide, jnp.floor_divide, args_maker)
     self._CompileAndCheck(jnp.floor_divide, args_maker)
 
+  @jtu.sample_product(
+      dtype=float_dtypes,
+      fn=["remainder", "mod", "floor_divide", "divmod"],
+  )
+  def test_float_zero_result_sign(self, fn, dtype):
```
Comment (jakevdp) [inline tests/lax_numpy_operators_test.py]: `test_divlike_result_sign` may be more descriptive. I'm imagining seeing a CI failure in `test_float_zero_result_sign` and being confused about what is actually being tested.

**Review by jakevdp — CHANGES_REQUESTED** [review]

File: tests/lax_numpy_operators_test.py (hunk 1 above)
Comment (kodlan) [reply tests/lax_numpy_operators_test.py]: done

File: tests/lax_numpy_operators_test.py (hunk 2 above)
Comment (kodlan) [reply tests/lax_numpy_operators_test.py]: done

**Review by jakevdp — APPROVED** [review]
Thanks!

**Comment by jakevdp** [comment]
The new test fails for float16 inputs on TPU v4, at the following locations:
```
Mismatched elements: 10 / 144 (6.94%)
First 5 mismatches are at indices:
 [125]: True (ACTUAL), False (DESIRED)
 [126]: True (ACTUAL), False (DESIRED)
 [127]: True (ACTUAL), False (DESIRED)
 [128]: True (ACTUAL), False (DESIRED)
 [129]: True (ACTUAL), False (DESIRED)
```
I suppose we could just skip the test on TPU backends?

**Comment by kodlan** [comment]
> The new test fails for float16 inputs on TPU v4, at the following locations:
> 
> ```
> Mismatched elements: 10 / 144 (6.94%)
> First 5 mismatches are at indices:
>  [125]: True (ACTUAL), False (DESIRED)
>  [126]: True (ACTUAL), False (DESIRED)
>  [127]: True (ACTUAL), False (DESIRED)
>  [128]: True (ACTUAL), False (DESIRED)
>  [129]: True (ACTUAL), False (DESIRED)
> ```
> 
> I suppose we could just skip the test on TPU backends?

done
